introduce store without file system details (#22)
nix-community/nix: doc/manual/src/architecture/store/store.md

                    |
                    V
             [ store object ]

<!-- -->

              [ reference ]
                    |
                    V
    [ store ] --> delete --> [ store' ]

It can *perform builds*, that is, create new store objects by transforming build inputs into build outputs, using instructions from the build tasks.

              [ reference ]
                    |
                    V
    [ store ] --> build
                       \
                      (maybe) --> [ store' ]
                         |
                         V
                   [ reference ]

As it keeps track of references, it can [garbage-collect][garbage-collection] unused store objects.

    [ store ] --> collect garbage --> [ store' ]

## Closure

Nix stores have the *closure property*: for each store object in the store, all the store objects it references must also be in the store.

Adding, building, copying and deleting store objects must be done in a way that obeys this property:

- A newly added store object cannot have references, unless it is a build task.

- Build results must only refer to store objects in the closure of the build inputs.

>   Building a store object will add appropriate references, according to provided build instructions.

Author: ```suggestion
  Building a store object will add appropriate references, according to the build task.
```

  These references can only come from declared build inputs.

- Store objects being copied must refer to objects already in the destination store.

  Recursive copying must either proceed in dependency order or be atomic.

- We can only safely delete store objects which are not reachable from any reference still in use.

>   Garbage collection will delete all store objects that cannot be reached from any reference in use.

Author: ```suggestion
  Garbage collection will delete those store objects that cannot be reached from any reference in use.
```

    file system object                          store path
            |                                        |
+-----------|----------------------------------------|------------+
| operating system        +------------+             |            |
|           '------------ |            | <-----------'            |
|                         |    file    |                          |
|                     ,-- |            | <-,                      |
|                     |   +------------+   |                      |
|          execute as |                    | read, write, execute |
|                     |   +------------+   |                      |
|                     '-> |  process   | --'                      |
|                         +------------+                          |
+-----------------------------------------------------------------+
```

Store objects are therefore implemented as the pair of

  - a *file system object* for data
  - a set of *store paths* for references.

There exist different types of stores, which all follow this model.
Examples:
- store on the local file system
- remote store accessible via SSH
- binary cache store accessible via HTTP

> Every store ultimately has to make store objects accessible to processes through the file system.

Author: ```suggestion
To make store objects accessible to processes, stores ultimately have to expose store objects through the file system.
```
